client/signup/config/flows.js

```diff
@@ -238,11 +250,84 @@ const Flows = {
 			flow = filterDesignTypeInFlow( flow );
 		}
 
-		return flow;
+		if ( ! user.get() ) {
+			// Preload the AB tests before going forward in the flow.
+			// Only valid for logged out users.
+			Flows.preloadABTestVariationsForStep( flowName, currentStepName );
+		}
+
+		return Flows.getABTestFilteredFlow( flowName, flow );
 	},
 
 	getFlows() {
 		return flows;
+	},
+
+	/**
+	 * Preload AB Test variations after a certain step has been completed.
+	 *
+	 * This gives the option to set the AB variation as late as possible in the
+	 * signup flow.
+	 *
+	 * Currently only the `main` flow is whitelisted.
+	 *
+	 * @param {String} flowName The current flow
+	 * @param {String} stepName The step that is being completed right now
+	 */
+	preloadABTestVariationsForStep( flowName, stepName ) {
+
+		/**
+		 * In cases where the flow is being resumed, the flow must not be changed from what the user
+		 * has seen before.
+		 *
+		 * E.g. A user is resuming signup from before the test was added. There is no need
+		 * to add a step somewhere back in the line.
+		 */
+		if ( Flows.resumingFlow ) {
+			return;
+		}
+
+		if ( 'main' === flowName ) {
+			if ( 'survey' === stepName ) {
+				abtest( 'siteTitleStep' );
+			}
+		}
+	},
+
+	/**
+	 * Return a flow that is modified according to the ABTest rules.
+	 *
+	 * Useful when testing new steps in the signup flows.
+	 *
+	 * Example usage: Inject or remove a step in the flow if a user is part of an ABTest.
+	 *
+	 * @param {String} flowName The current flow name
+	 * @param {Object} flow The flow object
+	 */
+	getABTestFilteredFlow( flowName, flow ){
+		/**
+		 * Filter according to running ABTests
+		 */
+
+		// Only do this on the main flow
+		if ( 'main' === flowName ) {
+			const updatedFlow = cloneDeep( flow );
+			
+			if ( getABTestVariation( 'siteTitleStep' ) === 'showSiteTitleStep' ) {
+				Flows.insertStepIntoFlow( 'survey', 'site-title', updatedFlow );
+			}
+
+			return updatedFlow;
+		}
+
+		return flow;
+	},
+
+	insertStepIntoFlow( afterStep, stepName, flow ) {
+		if ( -1 === flow.steps.indexOf( stepName ) ) {
+			const afterStepIndex = flow.steps.indexOf( afterStep );
+			flow.steps.splice( afterStepIndex + 1, 0, stepName );
```

Comment:
Instead of using the A/B test variants to determine whether to add an extra step, could we use it to determine whether to remove an existing step? That way:

1. We don't need the `afterStep` parameter.
2. The order of the steps goes in the `flows` object defined earlier (declarative), 
3. We make it possible to add optional A/B test steps to the last position, if we ever need that in the future.

We could name the method `flowOmittingStep()` or something similar. 😆